content/transfer-hook.md

```diff
@@ -0,0 +1,2918 @@
+
+        is_writable: bool,
+
+    ) -> Result<Self, ProgramError> {
+
+        Ok(Self {
+
+            discriminator: 1,
+
+            address_config: Seed::pack_into_address_config(seeds)?,
+
+            is_signer: is_signer.into(),
+
+            is_writable: is_writable.into(),
+
+        })
+
+    }
+
+
+
+    /// Create an `ExtraAccountMeta` from a list of seed configurations, representing
+
+    /// a PDA for an external program
+
+    ///
+
+    /// This PDA belongs to a program elsewhere in the account list, rather
+
+    /// than the executing program. For a PDA on the executing program, use
+
+    /// `ExtraAccountMeta::new_with_seeds`.
+
+    pub fn new_external_pda_with_seeds(
+
+        program_index: u8,
+
+        seeds: &[Seed],
+
+        is_signer: bool,
+
+        is_writable: bool,
+
+    ) -> Result<Self, ProgramError> {
+
+        Ok(Self {
+
+            discriminator: program_index
+
+                .checked_add(U8_TOP_BIT)
+
+                .ok_or(AccountResolutionError::InvalidSeedConfig)?,
+
+            address_config: Seed::pack_into_address_config(seeds)?,
+
+            is_signer: is_signer.into(),
+
+            is_writable: is_writable.into(),
+
+        })
+
+    }
+
+```
+
+
+
+We have three methods for creating an `ExtraAccountMeta`:
+
+
+
+1. `ExtraAccountMeta::new_with_pubkey` - For any normal account ( Not a program account )
+
+2. `ExtraAccountMeta::new_with_seeds` - For a program account PDA from the calling transfer hook program 
+
+3. `ExtraAccountMeta::new_external_pda_with_seeds` - For a program account PDA from a different external program
+
+
+
+Now that we know the accounts we can store them in `extra_account_meta_list`. Let's talk about the `InitializeExtraAccountMetaList` instruction itself. For most implementations, it should simply just create the `extra_account_meta_list` account and load it up with any additional accounts it needs. 
+
+
+
+Let's take a look at a simple example where we'll initialize an `extra_account_meta_list` with two additional arbitrary accounts, `some_account` and a `pda_account`. The `initialize_extra_account_meta_list` function will do the following:
+
+
+
+1. Prepare the accounts that we need to store in the `extra_account_meta_list` account as a vector (we'll talk about that in-depth in a sec).
+
+2. Calculate the size and rent required to store the list of `ExtraAccountMetas`.
+
+3. Make a CPI to the System Program to create an account and set the Transfer Hook Program as the owner, and then initialize the account data to store the list of `ExtraAccountMetas`.
+
+
+
+```rust
+
+#[derive(Accounts)]
+
+pub struct InitializeExtraAccountMetaList<'info> {
+
+  #[account(mut)]
+
+  payer: Signer<'info>,
+
+
+
+  /// CHECK: ExtraAccountMetaList Account, must use these seeds
+
+  #[account(
+
+        mut,
+
+        seeds = [b"extra-account-metas", mint.key().as_ref()],
+
+        bump
+
+    )]
+
+  pub extra_account_meta_list: AccountInfo<'info>,
```

Comment:
@jamesrp13

Reply:
This doesn't appear to be specific to the rust snippets. Something must have gone wrong copying the file over at some point because there are extra newlines everywhere. Pushing a commit now that removes thousands of empty lines 😂